core/color.cpp

```diff
@@ -47,6 +47,18 @@ uint32_t Color::to_ARGB32() const {
 	return c;
 }
 
+uint32_t Color::to_ABGR32() const {
+	uint32_t c = (uint8_t)(a * 255);
+	c <<= 8;
+	c |= (uint8_t)(b * 255);
+	c <<= 8;
+	c |= (uint8_t)(g * 255);
+	c <<= 8;
+	c |= (uint8_t)(r * 255);
+
+	return c;
+}
+
```

Comment:
toger this conversion is not necessary and does not do what it says, it is actually doing RGBA, the thing is. nanosvg uses RGBA but stores the color in an unsigned integer, leaving the alpha at the end, and it actualy ignores what is there, the alpha is handled by the opacity attribute  in SVG. so you dont need this. i think you are over complicating the color conversion.

Reply:
in any case the function should be called to_RGBA32, and should not ignore alpha

Reply:
I tried with both directions. And made sure I understand how it needs to be stored.
I think I have to do it this way. It does not work otherwise.
I know that alpha is not part of the color but since it is reading from the right I need to add it to the avoid an offset. of course I could just set it to 0xFF or just shift by 16 but it is more versatile like that...

@djrm feel free to send what you think would work. Maybe I have overseen something.

Reply:
This is what NanoSvg does: AABBGGRR
8 bit alpha, 8 bit BlueChannel, 8 bit GreenChannel, 8 bit RedChannel

I hope you are right that the inversion is caused by unsigned/signed int.
Although I think that makes no difference

Reply:
@toger5 the function is correct, the name is not, the function is doing RGBA not ABGR

Reply:
you are overlooking this, the nanosvg color format is

```
R  G  B  A
FF FF FF FF
```

Reply:
maybe you are thinking about bits in reverse direction
a shift moves the bits towards the left so `1 << 2 = 10`.

if you analize your function well, you will see

Reply:
if you are still not convinced this is from nanosvg source code
```
{ "red", NSVG_RGB(255, 0, 0) },
{ "green", NSVG_RGB( 0, 128, 0) },
{ "blue", NSVG_RGB( 0, 0, 255) },
```

Reply:
@toger5 sorry i misunderstood, i didnt see this
```
#define NSVG_RGB(r, g, b) (((unsigned int)r) | ((unsigned int)g << 8) | ((unsigned int)b << 16))
```

Reply:
@toger5 i think you can use the same formula, it is simpler

Reply:
like
```
 ((uint32_t)(r * 255)) | ((uint32t)(g * 255) << 8) | ((uint32t)(b * 255) << 16) | ((uint32t)(a * 255) << 24)
```

Reply:
I wrote some unnecessary long text... which is probably obvious for you anyways. 
but it proves the order of the nsvg color.

I also ask if the change form u_int to int could have caused the issue?
```cpp
return Color::hex(
	(p_hex & 0xFF) << 24 | //red to << 24
	(p_hex & (0xFF << 8)) << 8 | //green to << 16
	(p_hex & (0xFF << 16)) >> 8 | //blue to << 8
	0xFF //alpha to << 0 and set to 255
);
```
I think this is the funciton we are talking about.

p_hex is the value I get from nano svg.
I take the last unshifted 8 bits (looking from left to right). By doing the & opertation.
0xFF is 
0000 0000 0000 0000 0000 0000 1111 1111 (i think)
and the color maybe is: 
        1111 1111 0111 0001 1100 0110 0000 1010
so we get 
        0000 0000 0000 0000 0000 0000 0000 1010
than i shift it by 24 to the left (as you said this increases the int value)

this is the result:
        0000 1010 0000 0000 0000 0000 0000 0000
so we moved the red channel from the end to the beginning

But I honestly dont know how signed/unsigned int are stored. if that would mean we start a number in the center:
        0000 0000 0000 0001 0000 0000 0000 0000 = 1
        0000 0000 0000 0000 1000 0000 0000 0000 = -1
than this might be the reason why I had trouble because im assigning an uint to an int which would result in looping the bytes by 16...
If there negative values work by just using one bit at the front as signing, than that wouldnt really change much.

Reply:
this is literally the same:
```cpp
	uint32_t c = (uint8_t)(a * 255);
	c <<= 8;
	c |= (uint8_t)(b * 255);
	c <<= 8;
	c |= (uint8_t)(g * 255);
	c <<= 8;
	c |= (uint8_t)(r * 255);
```
I just used the same style than the other color conversion function for consistency

Reply:
> @toger5 sorry i misunderstood, i didnt see this

no problem at all. It took me sooo long to finally understand why e0e0e0 worked but all the other colors did not.
(for e0e0e0 it does not matter which oder...)

Reply:
I'm not an expert in bit shifting, but this ABGR really sounds like we're reading it wrong...

Reply:
It depends on the endianess of the system. In little endian systems this code produces `RGBA` because the `A` part gets shifted to the highest position. On big-endian systems this will be `ABGR`. So for little endian (which end consumer computers are) systems this code is wrong.

Reply:
The more system independent solution would be this

```c
union {
    uint8_t color_values[4];
    uint32_t compound_value;
} converter;

converter.color_values[0] = a;
converter.color_values[1] = b;
converter.color_values[2] = g;
converter.color_values[3] = r;

return converter.compound_value;
```

This unsures that the `A` part is always on the lowest address.

Reply:
It seems like nanosvg doesn't really care about memory layout, they do the encoding anyway, so the intial code looks okay to me.

Reply:
Alright, I played with my calculator a bit and I confirm that the code in this commit is equivalent to nanosvg's `(r << 0) | (g << 8) | (b << 16) | (a << 24)`.

For some turquoise color (R = 0, G = 128, B = 128, A = 255) we'd thus have:
```
(0 << 0) | (128 << 8) | (128 << 16) | (255 << 24)
= 1111 1111 1000 0000 1000 0000 0000 0000
= AA        GG        BB        RR
```
So yeah, if you read the memory byte by byte, it's indeed AGBR, but it's really just reading the int backwards IMO :)

Reply:
BTW, `Color::to_32()` and `Color::to_ARGB32()` are exactly the same no?